### PR TITLE
Add default version support

### LINKED ISSUE #1581: Support Default Version APIs
### Describe your problem(s)
Support default versioned apis where the version is not appended to the basePath.

### Describe your solution
Default versioned apis can be recognized from the api.yaml. Once it is identified, two routes should be created (with and without the version being appended to the envoy routes). When there is another version deployed and that version become the default version, previously created default version route should be replaced with the new version.

The enforcer-context parameters should remain versioned in both routes. When populating analytics, and rate limiting it should publish the events in a similar manner. 

### How will you implement it

---

### PR DESCRIPTION
**Pull request by praminda**
### Purpose

$subject.

default version change event is coming as an API_UPDATE type event from CP. This event doesn't contain enough information for CC to figureout if it should redeploy the API. Hence the `Action` property is introduced to API_UPDATE event.
Also API_UPDATE event doesn't contain the `gatewayLabels` property, therefore we are using internal maps to find the deployed envs of an API when default version change event is received.
### Issues

Fixes #1581

### Automation tests
 - Unit tests added: Yes
 - Integration tests added: Yes

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 1]
```diff
@@ -1322,3 +1325,8 @@ func getDefaultResourceMethods(apiType string) []string {
 	}
 	return defaultResourceMethods
 }
+
+func getDefaultVersionBasepath(basePath string, version string) string {
+	context := strings.ReplaceAll(basePath, "/"+version, "")
+	return "(" + basePath + "|" + context + ")"
```
Comment (renuka-fernando) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1331]:
```suggestion
	return fmt.Sprintf("(%s|%s)", basePath, context)
```

File: adapter/internal/messaging/notification_listener.go [hunk 2]
```diff
@@ -131,11 +132,36 @@ func processNotificationEvent(conf *config.Config, notification *msg.EventNotifi
...
+				query := make(map[string]string, 3)
+				query[eh.GatewayLabelParam] = configuredEnv
+				query[eh.ContextParam] = event.APIContext
+				query[eh.VersionParam] = event.APIVersion
+				eh.UpdatAPIMetadataFromCP(query)
+			}
+		}
+	}
+
+	synchronizer.FetchAPIsFromControlPlane(event.UUID, deployedEnvs)
```
Comment (praminda) [inline adapter/internal/messaging/notification_listener.go:156]: Add a comment on the flow for default version handling

File: adapter/internal/discovery/xds/server.go [hunk 3]
```diff
@@ -482,6 +482,18 @@ func GetAllEnvironments(apiUUID, vhost string, newEnvironments []string) []strin
 	return allEnvironments
 }
 
+// GetDeployedEnvironmets returns all the environments the API with `apiUUID` is deployed to
+func GetDeployedEnvironmets(apiUUID string) []string {
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/discovery/xds/server.go:486]:
```suggestion
func GetDeployedEnvironments(apiUUID string) []string {
```

File: adapter/internal/eventhub/subscription.go [hunk 4]
```diff
@@ -205,6 +205,15 @@ func LoadSubscriptionData(configFile *config.Config, initialAPIUUIDListMap map[s
 	go retrieveAPIListFromChannel(APIListChannel, nil)
 }
 
+// UpdatAPIMetadataFromCP Invokes `ApisEndpoint` and updates APIList synchronously.
+func UpdatAPIMetadataFromCP(params map[string]string) {
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/eventhub/subscription.go:209]:
```suggestion
func UpdateAPIMetadataFromCP(params map[string]string) {
```

File: adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go [hunk 5]
```diff
@@ -46,19 +46,16 @@ func TestGenerateRoutePaths(t *testing.T) {
 	basePath := "/basePath"
 	resourcePath := "/resource"
 
-	completeRoutePath := generateRoutePaths(xWso2BasePath, basePath, resourcePath)
+	filteredBasePath := getFilteredBasePath(xWso2BasePath, basePath)
+	completeRoutePath := generateRoutePath(filteredBasePath, resourcePath)
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/oasparser/envoyconf/envoyconf_internal_test.go:50]: I guess we can add a unit test to route path when default version is given. maybe in routes_with_clusters_test.go

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/DefaultVersionApiTestCase.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        Map<String, String> headers = new HashMap<>();
+        headers.put("Internal-Key", testKeyV2);
+        HttpResponse response = HttpsClientRequest.doGet("https://localhost:9095" + v2Context + "/store/inventory", headers);
+        Assert.assertNotNull(response);
+        Assert.assertEquals(response.getResponseCode(), HttpStatus.SC_OK, "Response code mismatched");
+        Assert.assertTrue(response.getData().contains("233539"), "Response body mismatched");
+    }
+
+    @Test(description = "Test invoking `non` default versioned API")
+    public void testInvokingNoneDefaultVersion() throws CCTestException {
```
Comment (AmaliMatharaarachchi) [inline integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/DefaultVersionApiTestCase.java:94]: shall we test auth token validation on default API request as well? I guess we have to add throttling and analytics tests as well.

Comment (praminda) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/apiDeploy/DefaultVersionApiTestCase.java:94]: You mean a bearer token validation or use v1 token on v2?

**Review by pubudu538 — APPROVED** [review]

**Pull request merged**